### ISSUE TITLE
Rule Change: no-param-reassign shouldn't stress over reassigning to destructured paramaters

### ISSUE THREAD

**Issue by cha0s** [eslint/eslint#15692]
### What rule do you want to change?

no-param-reassign

### What change to do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```js
const someFunction = ({foo, bar}) => {
  let something;
  if (!foo) {
    // Not actually mutating `arguments`.
    [something, foo] = whatever();
  }
}
```

### What does the rule currently do for this code?

It reports

> Assignment to function parameter 'foo'.

`foo` is not a function parameter.

### What will the rule do after it's changed?

The rule should not complain about reassigning a destructured parameter. Mutating `foo` is not actually mutating `arguments`, which is purportedly the reason this is lint in the first place:

![image](https://user-images.githubusercontent.com/298123/157729597-cdcaae58-9147-4c3c-b93b-05c9bf737c56.png)

### Participation

- [ ] I am willing to submit a pull request to implement this change.

**Comment by mdjermanovic** [comment]
Hi @cha0s, thanks for the issue!

This was already discussed in https://github.com/eslint/eslint/issues/8283. As stated in https://github.com/eslint/eslint/issues/8283#issuecomment-287567983 and [documentation](https://eslint.org/docs/rules/no-param-reassign) for this rule ("Often, assignment to function parameters is unintended and indicative of a mistake or programmer error."), this rule is not just about `arguments`, and it intentionally disallows assignments to any variables from the parameter list. Since `arguments` are rarely used in ES6+, this rule is mostly stylistic. If you want to allow assignment to some variables from the parameter list, you can disable this rule.
